src/util/color/rgbcolor.h

```diff
@@ -0,0 +1,153 @@
+// Apart from the assignment operator this type is immutable.
+class RgbColor {
+  public:
+    // A pure 24-bit 0xxRRGGBB color code with 8-bit per channel
+    // without an an alpha channel.
+    // We are using a separate typedef, because QRgb implicitly
+    // includes an alpha channel whereas this type does not!
+    typedef quint32 code_t;
+
+    static constexpr code_t validateCode(code_t code) {
+        return code & kRgbCodeMask;
+    }
+    static bool isValidCode(code_t code) {
+        return code == validateCode(code);
+    }
+
+    // The default constructor is not available, because there is
+    // no common default value that fits all possible use cases!
+    RgbColor() = delete;
+    // Explicit conversion from code_t with implicit validation.
+    explicit constexpr RgbColor(code_t code)
+            : m_code(validateCode(code)) {
+    }
+    // Explicit conversion from QColor.
+    RgbColor(QColor anyColor, code_t codeIfInvalid)
+            : m_code(anyColorToCode(anyColor, codeIfInvalid)) {
+    }
+
+    // Implicit conversion to a color code.
+    constexpr operator code_t() const {
+        return m_code;
+    }
+
+    friend bool operator==(RgbColor lhs, RgbColor rhs) {
+        return lhs.m_code == rhs.m_code;
+    }
+
+    typedef std::optional<RgbColor> optional_t;
+
+    static constexpr optional_t nullopt() {
+        return std::nullopt;
+    }
+
+    // Overloaded conversion functions for conveniently creating
+    // std::optional<RgbColor>.
+    static constexpr optional_t optional(code_t code) {
+        return optional(RgbColor(code));
+    }
+    static constexpr optional_t optional(RgbColor color) {
+        return std::make_optional(color);
+    }
+    static optional_t optional(QColor color) {
+        if (color.isValid()) {
+            return optional(validateCode(color.rgb()));
+        } else {
+            return nullopt();
+        }
+    }
+    static optional_t optional(const QVariant& varCode) {
+        if (varCode.isNull()) {
+            return nullopt();
+        } else {
+            DEBUG_ASSERT(varCode.canConvert(QMetaType::UInt));
+            bool ok = false;
+            const auto code = varCode.toUInt(&ok);
+            VERIFY_OR_DEBUG_ASSERT(ok) {
+                return nullopt();
+            }
+            return optional(static_cast<code_t>(code));
+        }
+    }
+
+  protected:
+    // Bitmask of valid codes = 0x00RRGGBB
+    static constexpr code_t kRgbCodeMask = 0x00FFFFFF;
+
+    static code_t anyColorToCode(QColor anyColor, code_t codeIfInvalid) {
+        if (anyColor.isValid()) {
+            // Strip alpha channel bits!
+            return validateCode(anyColor.rgb());
+        } else {
+            return codeIfInvalid;
+        }
+    }
+
+    code_t m_code;
+};
+
+inline bool operator!=(RgbColor lhs, RgbColor rhs) {
+    return !(lhs == rhs);
+}
+
+// Explicit conversion of both non-optional and optional
+// RgbColor values to QColor as overloaded free functions.
```

Comment:
Why you  decide her for a free function? I think it is more QT like to have a member function for this.

Reply:
We cannot add new functions to `std::optional`. The conversion functions for `RgbColor` and `std::optional<RgbColor` should use the same invocation pattern. In C++ this is only  possible with a free function.

Reply:
Why is std::optional relevant here?
Can't we just introduce
RgbColor::toQColor()

Reply:
No. Because std::nullopt -> QColor() (= invalid) is not cover by `RgbColor`.

Reply:
Sorry, I don't understand the point here.
I think I have missed something.

We have here the toQColor() function taking a RgbColor. There is no std::nullopt involved.

```
RgbColor rgb
QColor c1 = toQColor(rgb);
QColor c2 = rgb.toQColor();
```
What is the issue with the later?

Reply:
```
RgbColor rgb1 = ...;
std::optional<RgbColor> rgb2 = ...;
QColor col1 = toQColor(rgb1);
QColor col2 = toQColor(rgb2);
```

Reply:
toQColor here looks like a constructor, can it be one of those instead?

auto col1 = QColor(rgb1);

...

explicit QColor(RgbColor rgb) {};

Reply:
and explicit QColor(std::optiona<RgbColor> rgb_or)  {}

Reply:
How could you add an additional constructor to `QColor`??

Reply:
erp my bad, I thought QColor was our class.

Reply:
oh I have some ideas, hit me up on zulip

Reply:
ok sounds like we have consensus on this PR!

Reply:
The decision to map `std::nullopt` to `QColor()` should be done here, only once and it should be tested. Instead of hoping that developers use this convention at various places.

Reply:
DRY

Reply:
The same arguments apply for the conversion to `QVariant`. Let's not repeat those bad practices in the current code base again and again.